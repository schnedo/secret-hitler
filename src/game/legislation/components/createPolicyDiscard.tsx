import type { ComponentType, ReactElement } from "react";
import type { ModalProps } from "../../../components";
import type PolicyDeck from "../PolicyDeck";

export interface DiscardPolicyProps {
  onDiscard: (index: number) => void;
  onPlay: () => void;
  isChancellorDiscard?: boolean;
  drawingPile: PolicyDeck;
}

export default function createPolicyDiscard(Modal: ComponentType<ModalProps>) {
  return function DiscardPolicy({
    onDiscard,
    onPlay,
    isChancellorDiscard,
    drawingPile,
  }: DiscardPolicyProps): ReactElement {
    const nCardsToDraw = isChancellorDiscard ? 2 : 3;
    const handleClick = (index: number) => {
      onDiscard(index);
      if (isChancellorDiscard) {
        onPlay();
      }
    };

    return (
      <Modal open>
        <div>select a policy to discard</div>
        {drawingPile.slice(0, nCardsToDraw).map((policy, index) => (
          <button key={index} onClick={() => handleClick(index)}>
            {policy}
          </button>
        ))}
      </Modal>
    );
  };
}
