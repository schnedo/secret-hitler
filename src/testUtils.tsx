import type { PropsWithChildren } from "react";

export function MockComponent({
  children,
  ...props
}: PropsWithChildren<unknown>): JSX.Element {
  return (
    <div>
      {JSON.stringify(props)} {children}
    </div>
  );
}
