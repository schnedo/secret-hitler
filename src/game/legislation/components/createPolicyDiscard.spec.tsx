import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { MockComponent } from "../../../testUtils";
import type PolicyDeck from "../PolicyDeck";
import createPolicyDiscard from "./createPolicyDiscard";

const PolicyDiscard = createPolicyDiscard(MockComponent);

const drawingPile: PolicyDeck = ["fascist", "liberal", "fascist"];

it("should render correctly", async () => {
  expect.hasAssertions();

  const { container } = render(
    <PolicyDiscard
      onDiscard={() => undefined}
      onPlay={() => undefined}
      drawingPile={drawingPile}
    />,
  );

  expect(container).toMatchSnapshot();
});

it("should draw 3 cards for president discard", async () => {
  expect.hasAssertions();

  render(
    <PolicyDiscard
      onDiscard={() => undefined}
      onPlay={() => undefined}
      drawingPile={drawingPile}
    />,
  );

  const allCards = screen.getAllByRole("button");
  expect(allCards).toHaveLength(3);
  const cardTypes = allCards.map((card) => card.textContent);
  expect(cardTypes).toStrictEqual(drawingPile.slice(0, 3));
});

it("should draw 2 cards for chancellor discard", async () => {
  expect.hasAssertions();

  render(
    <PolicyDiscard
      onDiscard={() => undefined}
      onPlay={() => undefined}
      drawingPile={drawingPile}
      isChancellorDiscard
    />,
  );

  const allCards = screen.getAllByRole("button");
  expect(allCards).toHaveLength(2);
  const cardTypes = allCards.map((card) => card.textContent);
  expect(cardTypes).toStrictEqual(drawingPile.slice(0, 2));
});

it("should call onDiscard when card is discarded", async () => {
  expect.hasAssertions();

  const handleDiscard = jest.fn();
  render(
    <PolicyDiscard
      onDiscard={handleDiscard}
      onPlay={() => undefined}
      drawingPile={drawingPile}
    />,
  );

  expect(handleDiscard).not.toHaveBeenCalled();
  await userEvent.click(screen.getByRole("button", { name: "liberal" }));
  expect(handleDiscard).toHaveBeenCalledTimes(1);
  expect(handleDiscard).toHaveBeenLastCalledWith(1);
});

it("should call onPlay when card is discarded by chancellor", async () => {
  expect.hasAssertions();

  const handlePlay = jest.fn();
  render(
    <PolicyDiscard
      onDiscard={() => undefined}
      onPlay={handlePlay}
      drawingPile={drawingPile}
      isChancellorDiscard
    />,
  );

  expect(handlePlay).not.toHaveBeenCalled();
  await userEvent.click(screen.getAllByRole("button").pop()!);
  expect(handlePlay).toHaveBeenCalledTimes(1);
});
