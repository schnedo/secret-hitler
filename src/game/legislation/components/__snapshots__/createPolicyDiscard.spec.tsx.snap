// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div>
  <div>
    {"open":true}
     
    <div>
      select a policy to discard
    </div>
    <button>
      fascist
    </button>
    <button>
      liberal
    </button>
    <button>
      fascist
    </button>
  </div>
</div>
`;
